contribute/design/1.x/upgrade/volume-pools-upgrade.md

```diff
@@ -0,0 +1,499 @@
+
+    For example, a cstor-operator reads a user generated resource
+    called SPC, as well as creates a resource called CSP. It is possible
+    that both of these user-facing and internal resources go through
+    a change to support some new functionality. With this design, 
+    when the cstor operator is upgraded to latest version, the following
+    will happen:
+    - The non-user facing changes in SPC are handled by the cstor 
+      operator, which could be like adding a finalizer or some 
+      annotation that will help with internal bookkeeping. It will
+      also continue to read and operate on the existing schema. If
+      there has been a user api change, the steps will be provided 
+      for the administrator to make the appropriate changes to their
+      SPC (which is probably saved in their private Git) and re-apply 
+      the YAML. 
+    - The changes to the CSP will be managed completely by the upgraded
+      cstor operator. User/Administrator need not interact with CSP. 
+
+    Note: This eliminates the need for pre-upgrade scripts that were used 
+    till 1.0. 
+
+    Status: Available in 1.1
+
+2.  Data plane components have an interdependence on each other and
+    upgrading a volume typically involves upgrading multiple related
+    components at once. For example upgrading a jiva volume involves
+    upgrading the target and replica deployments. The functionality 
+    to upgrade these various components are provided and delivered
+    in a container hosted at `quay.openebs.io/m-upgrade`. 
+
+    A Kubernetes Job with this upgrade container can be launched to 
+    carry out the upgrade process. The upgrade job will have to run
+    in the same namespace as openebs control plane and with same
+    service account.
+
+    Upgrade of each resource will be launched as a separate Job, that
+    will allow for parallel upgrading of resources as well as 
+    granular control on which resource is being upgraded. Upgrades
+    are almost seamless, but still have to be planned to be run 
+    at different intervals for different applications to minimize
+    the impact of unforeseen events.
+
+    As the upgrades are done via Kubernetes job, a pod is scheduled
+    to perform the task within the cluster, eliminating any network 
+    dependencies that might there between machine that triggers
+    the upgrade to cluster. Also the logs of the upgrade are saved
+    on the pod. 
+
+    Note: This replaces the script based upgrades from OpenEBS 1.0. 
+    Sample Kubernetes YAMLs for upgrading various resources can be
+    found [here](../../../../k8s/upgrades/1.0.0-1.1.0/).  
+
+    Status: Available in 1.1 and supports upgrading of Jiva Volumes,
+    cStor Volumes and cStor pools from 1.0 to 1.1
+
+3.  A custom resource called `UpgradeTask` will be defined with the 
+    details of the object being upgraded, and will in turn be updated
+    with the status of the upgrade by the upgrade-container. This 
```

Comment:
If this custom resource deals only with history or status, then we might want to rename accordingly.

Reply:
Will this custom resource i.e. UpgradeTask be used to continue from the failed point?

Reply:
It will be used both as input and output. updated the text. 

The second question is interesting. The upgrade Job can be re-attempted and it will start off from where it left off previously. I will update this in the UpgradeJob section.